debian/wviewhtmlconfig.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH WVIEWHTMLCONFIG 1 "November 19, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
wviewhtmlconfig \- script to configure the HTML templates for a wview installation
.SH SYNOPSIS
.B wviewhtmlconfig
.RI
.br
.SH DESCRIPTION
This manual page describes the use of
.B wviewhtmlconfig
.
Configure a wview installation's HTML templates.
.BR
.BR Note: wviewconfig MUST be run before this script.
.BR Note: must be executed on the source wview server.
.SH OPTIONS
There are no options.
.TP
.SH SEE ALSO
.BR wviewconfig (1),
.BR /usr/share/doc/wview/wview-User-Manual.html
.br
.SH AUTHOR
wviewhtmlconfig was written by Mark Teel.
.PP
This manual page was written by Mark Teel <mteel2005@gmail.com>,
for the Debian project (and may be used by others).


ChangeLog
RELEASE VERSION HISTORY (newest to oldest)
------------------------------------------


5.21.5    04-06-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Remove debug message.




5.21.4    04-06-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Added stale pid file cleanup to the APT package start script.

2)  WMRUSB HID interface changes. Changed reader thread to blocking IO.
    Mitigated the console occasionally blasting out bogus data. Now zero
    packet loss. Fixed issue with pressure sanity check. Added sanity
    check for bogus rain accumulator values from the WMR console.




5.21.3    03-31-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Further WMRUSB interface improvements. Better packet framing.




5.21.2    03-30-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Updated debian APT postinst script to include copying the exfoliation skin
    for new installs.

2)  Enabled checksum processing for WMR stations.

3)  Changed LOOP windspeed items to floats for better precision.
    NOTE: If you use custom datafeed clients the LOOP data structure has changed.

4)  Added temp only packet support for WMRUSB stations. This adds support for
    add-on sensors such as the pool sensor.

5)  Fixed APT installs changing the wview.log file ownership incorrectly. Also
    restarts rsyslog automatically if a new install.




5.21.1    03-13-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Modified HILOW database storage so higher precision is used for cumulative
    data such as rain and ET.

2)  Fixed issue with yearly historical charts when generated at wview start
    (as opposed to at midnight). The current day was erroneously being included in
    the history data. This was most apparent with the UV and Solar Radiation
    charts.

3)  Fixed bug introduced in 5.21.0 release causing NULL wind direction to have
    360 added to it (to normalize calibration result). A value of -99640 was
    being stored/displayed.




5.21.0    03-05-2014
--------------------

radlib version required: 2.12.0 or newer

1)  WMRUSB improvements. Further reverse engineering of the undocumented protocol. Reset and
    heartbeat messages now working. Packet reception more stable. Initial packet acquisition
    faster resulting in faster startup.

2)  Modify CWOP packet header for new stricter compliance requirements of CWOP servers.

3)  Fix log message time_t bug for 64-bit platforms.

4)  Add timeout parameter for ssh rules. Added rows to config database.

5)  Updated data feed conversion logic to handle ARCHIVE_VALUE_NULL in loop data.

6)  Added flat line sensor value detection for outside temp, wind direction and wind speed.
    If there are no changes over a 4 hour period an email alert will be sent.

7)  Added Station Name, City and State to the wviewmgmt System Status page header.

8)  Added logic to SSH timer to fixup next timer interval if system time has skewed or
    if the ssh transfer took longer than 2 minutes.

9)  Fixed TE923 rain calculation to assume each tip is 1/36" instead of 1mm.

10) Fixed a bug for Vantage Pro stations which could cause the last archive interval of
    the previous day to be included in the current day HILOW data. This occurred if the
    midnight archive record was provided late by the console.

11) Added wet bulb temperature calculation and html tag --WetBulbTemp--. Added to all
    "readings.incx" templates.

12) Added email alert notification if no loop data has been received for a given archive
    interval. This applies to all stations except the Davis stations (alerts already
    exist for them).

13) Updated start scripts to detect and delete stale pid files.

14) Modified WH1080 data parsing to assume a humidity of 9% if the value from the
    station is undefined (0xff).




5.20.2    03-21-2012
--------------------

radlib version required: 2.12.0 or newer

1)  Added back "dailyrainin" for http reporting. They do no computation based on hourly rain values.

2)  Fixed a typo in the wviewmgmt station.php file. 




5.20.1    03-20-2012
--------------------

radlib version required: 2.12.0 or newer

1)  Modified http rain submission (Wunderground and Weatherforyou) so only the
    "rainin" parameter is sent so daily rain will be correct. 




5.20.0    03-17-2012
--------------------

radlib version required: 2.12.0 or newer

1)  Added "wait_for_time_set" function to most wview start script examples so 
    wview daemons are not started until the system time is set. This applies to
    platforms without a real time clock which must reset their time during 
    startup.

2)  Updated the User Manual section on skin selection to better describe enabling
    extended data generation (UV, ET, radiation).

3)  Updated the debian post install script to try harder to find the default http
    document root location for creation of symbolic links to wviewmgmt and the
    weather home page.

4)  Added new configuration option to optionally disable using the Vantage Pro 
    console archive records and instead autogenerate them from the LOOP data.
    This is how all other stations work. It also supports non-standard console
    interfaces (those not using the datalogger on the console).

5)  Updated the Weatherforyou URL for data submissions.

6)  Fixed bug with DataFeed clients that caused an infinite loop if you
    restarted wview while there was a datafeed client connection.

7)  Fixed bug with Alarm Type configuration introduced in 5.19.0.

8)  Changed TWI station to compute rain fall rate as it is reported to not
    be provided by the station.

9)  Added socket retry logic to the station ethernet interface (used by the
    Vantage Pro IP datalogger and any serial-to-ethernet device).

10) Modified CWOP host name to IP address resolution so that 'getaddrinfo' is
    used instead of 'gethostbyname' and all results are tried in the order they 

wviewconfig/wviewhtmlconfig.sh
################################################################################
#
# File:           wviewhtmlconfig.sh
#
# Description:    Provide a script to interactively configure wview HTML 
#                 templates.
#
# Usage:          (must be root)
#                 wviewhtmlconfig <skin_selection>
#
# History:
# Engineer	  Date	    Ver   Comments
# MS Teel         02/05/08   1    Original
# MS Teel         09/13/08   2    Modify for SQLite configuration DB
# MS Teel         05/03/09   3    Add non-interactive support
# MS Teel         03/12/12   4    Add exfoliation skin.
#
################################################################################


################################################################################
#################################  M A C R O S  ################################
################################################################################


################################################################################
#######################  D E F I N E  F U N C T I O N S  #######################
################################################################################

get_wviewconfig()
{
    echo "select value from config where name='HTMLGEN_EXTENDED_DATA';" > $WVIEW_CONF_DIR/cmnd.sql
    echo ".read $WVIEW_CONF_DIR/cmnd.sql" | sqlite3 $WVIEW_CONF_DIR/wview-conf.sdb > $WVIEW_CONF_DIR/value.out
    DATA_EXTENDED=`cat $WVIEW_CONF_DIR/value.out`

    echo "select value from config where name='HTMLGEN_METRIC_UNITS';" > $WVIEW_CONF_DIR/cmnd.sql
    echo ".read $WVIEW_CONF_DIR/cmnd.sql" | sqlite3 $WVIEW_CONF_DIR/wview-conf.sdb > $WVIEW_CONF_DIR/value.out
    STATION_METRIC=`cat $WVIEW_CONF_DIR/value.out`

    echo "select value from config where name='HTMLGEN_IMAGE_PATH';" > $WVIEW_CONF_DIR/cmnd.sql
    echo ".read $WVIEW_CONF_DIR/cmnd.sql" | sqlite3 $WVIEW_CONF_DIR/wview-conf.sdb > $WVIEW_CONF_DIR/value.out
    HTML_DESTINATION=`cat $WVIEW_CONF_DIR/value.out`

    rm -rf $WVIEW_CONF_DIR/cmnd.sql
    rm -rf $WVIEW_CONF_DIR/value.out
}

show_skins()
{
    echo ""
    echo "wview site skins currently supported:"
    echo "    classic - default wview skin"
    echo "    chrome  - classic with a chrome effect"
    echo "    exfoliation - a clean implementation by Matthew Wall"
}

interactive_intro()
{
    echo "################################################################################"
    echo " !!!!!!!!!!!!!!!!         READ THIS BEFORE PROCEEDING         !!!!!!!!!!!!!!!!"
    echo ""
    echo "--> HTML Template Configuration for wview"
    echo ""
    echo "--> Note: This script will save existing templates to $WVIEW_CONF_DIR/html-YYYYMMDD.HHmmSS"
    echo "          before writing the new files based on your answers here - if that"
    echo "          is not what you want, hit CTRL-C now to abort this script!"
    echo ""
    echo "################################################################################"
    echo ""
    echo -n "pausing 5 seconds "
    sleep 1
    echo -n "."
    sleep 1
    echo -n "."
    sleep 1
    echo -n "."
    sleep 1
    echo -n "."
    sleep 1
    echo "."
    echo ""
}

save_old()
{
    OLD_HTML_SAVE=`date '+%Y%m%d.%H%M%S'`
    echo "Saving old HTML directory to html-$OLD_HTML_SAVE ..."
    cp -rf $WVIEW_CONF_DIR/html $WVIEW_CONF_DIR/html-$OLD_HTML_SAVE
    echo "...done."
    echo "Saving old config files..."
    if [ -f $WVIEW_CONF_DIR/html-templates.conf ]; then
        mv $WVIEW_CONF_DIR/html-templates.conf $WVIEW_CONF_DIR/html-templates.conf-$OLD_HTML_SAVE
    fi
    if [ -f $WVIEW_CONF_DIR/images.conf ]; then
        mv $WVIEW_CONF_DIR/images.conf $WVIEW_CONF_DIR/images.conf-$OLD_HTML_SAVE
    fi
    if [ -f $WVIEW_CONF_DIR/images-user.conf ]; then
        mv $WVIEW_CONF_DIR/images-user.conf $WVIEW_CONF_DIR/images-user.conf-$OLD_HTML_SAVE
    fi
    if [ -f $WVIEW_CONF_DIR/graphics.conf ]; then
        mv $WVIEW_CONF_DIR/graphics.conf $WVIEW_CONF_DIR/graphics.conf-$OLD_HTML_SAVE
    fi
    if [ -f $WVIEW_CONF_DIR/pre-generate.sh ]; then
        mv $WVIEW_CONF_DIR/pre-generate.sh $WVIEW_CONF_DIR/pre-generate.sh-$OLD_HTML_SAVE
    fi
    if [ -f $WVIEW_CONF_DIR/post-generate.sh ]; then
        mv $WVIEW_CONF_DIR/post-generate.sh $WVIEW_CONF_DIR/post-generate.sh-$OLD_HTML_SAVE
    fi
    echo "... done."
}

remove_all()
{
	rm -f $WVIEW_CONF_DIR/html/*.*
	rm -f $WVIEW_CONF_DIR/html-templates.conf
	rm -f $WVIEW_CONF_DIR/images.conf
	rm -f $WVIEW_CONF_DIR/images-user.conf
	rm -f $WVIEW_CONF_DIR/graphics.conf
	rm -f $WVIEW_CONF_DIR/pre-generate.sh
	rm -f $WVIEW_CONF_DIR/post-generate.sh    
}

get_user_preferences()
{
    echo ""
    echo "---------------------------------------------------------------------------" 
    echo "Which template skin do you want to use for your site?"
    show_skins
    echo -n "($SITE_SKIN): "
    read INVAL
    if [ "" != "$INVAL" ]; then
        SITE_SKIN=$INVAL
    fi
    case "$SITE_SKIN" in
        "classic" )
            echo "Site skin $SITE_SKIN selected..."
            ;;
        "chrome" )
            echo "Site skin $SITE_SKIN selected..."
            ;;
        "exfoliation" )
            echo "Site skin $SITE_SKIN selected..."
            ;;
        *)
            echo "$SITE_SKIN not supported!"
            show_skins
            echo "Aborting - run wviewhtmlconfig again and select a valid skin..."
            exit 1
            ;;
    esac

    if [ "yes" = "$DATA_EXTENDED" ]; then
        echo "With extended data graphics."
    else
        echo "With NO extended data graphics."
    fi
    if [ "yes" = "$STATION_METRIC" ]; then
        echo "With metric units."
    else
        echo "With US/Imperial units."
    fi
}

move_templates()
{
    if [ "yes" = "$DATA_EXTENDED" ]; then
        cp -rf $WVIEW_CONF_DIR/html/$SITE_SKIN/plus/*.* $WVIEW_CONF_DIR/html
    else
        cp -rf $WVIEW_CONF_DIR/html/$SITE_SKIN/standard/*.* $WVIEW_CONF_DIR/html
    fi

    if [ -d $WVIEW_CONF_DIR/html/$SITE_SKIN/static ]; then
        cp -rf $WVIEW_CONF_DIR/html/$SITE_SKIN/static/*.* $HTML_DESTINATION
    fi
}

setup_conf_files()
{
    cp -f $WVIEW_CONF_DIR/html/html-templates.conf $WVIEW_CONF_DIR
    cp -f $WVIEW_CONF_DIR/html/pre-generate.sh $WVIEW_CONF_DIR
	chmod +x $WVIEW_CONF_DIR/pre-generate.sh
    cp -f $WVIEW_CONF_DIR/html/post-generate.sh $WVIEW_CONF_DIR
	chmod +x $WVIEW_CONF_DIR/post-generate.sh
    cp -f $WVIEW_CONF_DIR/html/images-user.conf $WVIEW_CONF_DIR
    cp -f $WVIEW_CONF_DIR/html/graphics.conf $WVIEW_CONF_DIR
    if [ "yes" = "$STATION_METRIC" ]; then
        cp -f $WVIEW_CONF_DIR/html/images-metric.conf $WVIEW_CONF_DIR/images.conf
        cp -f $WVIEW_CONF_DIR/html/awekas_wl.htx-metric $WVIEW_CONF_DIR/html/awekas_wl.htx
    else
        cp -f $WVIEW_CONF_DIR/html/images.conf $WVIEW_CONF_DIR/images.conf
        cp -f $WVIEW_CONF_DIR/html/awekas_wl.htx-us $WVIEW_CONF_DIR/html/awekas_wl.htx
    fi
}

################################################################################
##################  S C R I P T  E X E C U T I O N  S T A R T  #################
################################################################################
export PATH=/bin:/usr/bin:/sbin:/usr/sbin:/usr/local/bin:/usr/local/sbin

get_wviewconfig

#  Is there an argument?
if [ "" != "$1" ]; then
    if [ "$1" = "classic" ]; then
        SITE_SKIN=$1
        INTERACTIVE=0
    elif [ "$1" = "chrome" ]; then
        SITE_SKIN=$1
        INTERACTIVE=0
    elif [ "$1" = "exfoliation" ]; then
        SITE_SKIN=$1
        INTERACTIVE=0
    elif [ "$1" = "remove" ]; then
        REMOVE_IT=1
    else
        SITE_SKIN="chrome"
        INTERACTIVE=1
    fi
else
    REMOVE_IT=0
    SITE_SKIN="chrome"
    INTERACTIVE=1
fi

if [ "$REMOVE_IT" = "1" ]; then
    remove_all
    echo "HTML setup files removed"
    exit 0
fi

if [ "$INTERACTIVE" = "1" ]; then
    interactive_intro
    get_user_preferences
    save_old
fi

move_templates
setup_conf_files
echo "HTML setup complete:"
echo "you may now customize template files in $WVIEW_CONF_DIR/html"

